JDK-8179097: Fix NPE on removing mnemonics of disposed popup

vlaaad opened the pull request:
Hi! This is a follow-up PR for issue https://github.com/javafxports/openjdk-jfx/issues/465
I signed OCA (received welcome letter from @kevinrushforth on 26th of July).

I managed to create a stable repro case, see `MenuButtonSkinBaseNPETest.java`.
Most important lines in repro are 45-46:
```java
            menu.hide();
            menuBar.getMenus().clear();
```
When you hide menu which is in a menu bar, it scheduled mnemonics removal in runLater:
```java
Platform.runLater(() -> Utils.removeMnemonics(popup, getSkinnable().getScene()));
```
But when you clear said menu bar, it disposes all menus, which leaded to all sorts of exceptions: `getSkinnable` in MenuButtonSkinBase becomes null, but also popup's skin becomes null (and we access it later in `Utils::removeMnemonics`).

The fix is to save added mnemonics to a list so we don't depend on external state which will be mutated on disposing, and use that list when we want to remove them.

I run all tests with `bash ./gradlew -PFULL_TEST=true -PUSE_ROBOT=true -x web:test all test`.
Before fix:
```
649 tests completed, 20 failed, 119 skipped
```
One of the failed tests is what I added:
```
test.javafx.scene.control.MenuButtonSkinBaseNPETest > testMenuButtonNPE FAILED
    org.junit.ComparisonFailure: No error should be thrown expected:<[]> but was:<[Exception in thread "JavaFX Application Thread" java.lang.NullPointerException
        at javafx.controls/javafx.scene.control.skin.MenuButtonSkinBase.lambda$new$7(MenuButtonSkinBase.java:195)
        at javafx.graphics/com.sun.javafx.application.PlatformImpl.lambda$runLater$10(PlatformImpl.java:428)
        at java.base/java.security.AccessController.doPrivileged(Native Method)
        at javafx.graphics/com.sun.javafx.application.PlatformImpl.lambda$runLater$11(PlatformImpl.java:427)
        at javafx.graphics/com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatcher.java:96)
        at javafx.graphics/com.sun.glass.ui.gtk.GtkApplication._runLoop(Native Method)
        at javafx.graphics/com.sun.glass.ui.gtk.GtkApplication.lambda$runLoop$11(GtkApplication.java:277)
        at java.base/java.lang.Thread.run(Thread.java:834)
    ]>
        at org.junit.Assert.assertEquals(Assert.java:123)
        at test.javafx.scene.control.MenuButtonSkinBaseNPETest.testMenuButtonNPE(MenuButtonSkinBaseNPETest.java:50)
Exception in thread "JavaFX Application Thread" 
```
After fix:
```
649 tests completed, 17 failed, 119 skipped
```
My repro case is not present in the list of failed tests.

kevinrushforth commented on modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/skin/Utils.java:
```diff
@@ -693,6 +694,10 @@ public static void addMnemonics(ContextMenu popup, Scene scene) {
+        addMnemonics(popup, scene, initialState, null);
+    }
+
+    public static void addMnemonics(ContextMenu popup, Scene scene, boolean initialState, List<Mnemonic> into) {
```
This method must not be public (or protected), or it becomes part of the public API.

kevinrushforth commented:
I'll review this (more carefully this time), when I can. Have you ensured that there is no memory leak as a result of adding the list of mnemonics?

kevinrushforth commented on tests/system/src/test/java/test/javafx/scene/control/MenuButtonSkinBaseNPETest.java:
```diff
@@ -0,0 +1,84 @@
+package test.javafx.scene.control;
```
This file will need a copyright header block (you can copy one from another test using just `2019,` as the year.

vlaaad commented:
> Have you ensured that there is no memory leak as a result of adding the list of mnemonics?
No, I did not ensure that, although I clear the list once popup is hidden. Is it enough? How can I ensure that?

vlaaad commented:
@kevinrushforth I just realized my fix can misbehave in a scenario where we consecutively show, then hide, then show popup again: mnemonics list will be populated twice, and then cleared later in `runLater` while popup is still showing, and a proper fix would be making a copy of a mnemonics list that is then used in `runLater` while clearing original list in handler immediately.

Do you have any other issues with this PR that I should address, besides copyright header?

kevinrushforth commented:
> Do you have any other issues with this PR that I should address, besides copyright header?

I haven't had time to review it, so nothing else for now.

vlaaad commented:
Added license header, fixed mnemonics clearing so that we keep mnemonics list in sync with popup's showing property, and yet ensure that proper mnemonics are removed from scene in `runLater`

kevinrushforth commented:
I see one new test failure on the Mac that you will need to take a look at.

FYI, It may be next week before I can get back to this.

kevinrushforth commented:
Pointer to JBS issue: https://bugs.openjdk.java.net/browse/JDK-8179097

kevinrushforth commented:
Have you looked at the failure on Mac yet? You can click on the "details" link for the failing build in the section below.

Here is the failure I see:

```
test.javafx.scene.control.TextInputControlTest > UndoRedoTimestampBased[1] FAILED
    java.lang.AssertionError: 
        at org.junit.Assert.fail(Assert.java:91)
        at org.junit.Assert.assertTrue(Assert.java:43)
        at org.junit.Assert.assertTrue(Assert.java:54)
        at test.javafx.scene.control.TextInputControlTest.UndoRedoTimestampBased(TextInputControlTest.java:2066)
```

vlaaad commented:
I'm on vacation right now, will have a look next week when I'll be back

vlaaad commented:
@kevinrushforth I had a look and couldn't find anything related to my changes at all. Maybe this test just can fail once in a blue moon? I tried updating this branch to latest develop, and now tests pass.

kevinrushforth commented:
It was probably an intermittent failure then (although I've never seen that particular test fail). In any case, I'll put this on my review queue now.

kevinrushforth commented on tests/system/src/test/java/test/javafx/scene/control/MenuButtonSkinBaseNPETest.java:
```diff
@@ -0,0 +1,109 @@
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        System.setErr(new PrintStream(out,true));
+        Thread.sleep(1000);
+        Platform.runLater(()-> {
```
This should probably use `Util.runAndWait` instead. As it is, there is no guarantee that the runnable has completed before you check for the error being written to `System.err`.

kevinrushforth commented on tests/system/src/test/java/test/javafx/scene/control/MenuButtonSkinBaseNPETest.java:
```diff
@@ -0,0 +1,109 @@
+            menu.hide();
+            menuBar.getMenus().clear();
+        });
+        Thread.sleep(100);
```
You will still need this sleep even with the `runAndWait`.

kevinrushforth commented on tests/system/src/test/java/test/javafx/scene/control/MenuButtonSkinBaseNPETest.java:
```diff
@@ -0,0 +1,109 @@
+        });
+        Thread.sleep(100);
+        System.setErr(defaultErrorStream);
+        Assert.assertEquals("No error should be thrown","", out.toString());
```
Minor: please add a space after the first comma.

kevinrushforth commented on tests/system/src/test/java/test/javafx/scene/control/MenuButtonSkinBaseNPETest.java:
```diff
@@ -0,0 +1,109 @@
+            }
+        } catch (InterruptedException ex) {
+            Assert.fail("Unexpected exception: " + ex);
+        }
```
If you add `throws Exception` to the method, you can eliminate the try...catch block, and simplify it to the following, which we've done in other more recently-written tests:

```
    assertTrue("Timeout waiting for FX runtime to start", startupLatch.await(15, TimeUnit.SECONDS));
```

vlaaad commented:
Adressed Kevin's review concerns.

kevinrushforth commented on tests/system/src/test/java/test/javafx/scene/control/MenuButtonSkinBaseNPETest.java:
```diff
@@ -0,0 +1,105 @@
+    public void testMenuButtonNPE() throws Exception {
+        PrintStream defaultErrorStream = System.err;
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        System.setErr(new PrintStream(out,true));
```
Minor: add space after comma.

kevinrushforth commented on tests/system/src/test/java/test/javafx/scene/control/MenuButtonSkinBaseNPETest.java:
```diff
@@ -0,0 +1,105 @@
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        System.setErr(new PrintStream(out,true));
+        Thread.sleep(1000);
+        Util.runAndWait(()-> {
```
Minor: add space between `()` and `->`

vlaaad commented:
Fixed!

vlaaad commented on modules/javafx.controls/src/main/java/javafx/scene/control/skin/MenuButtonSkinBase.java:
```diff
@@ -185,7 +192,10 @@ public MenuButtonSkinBase(final C control) {
                 // consumed to prevent them being used elsewhere).
                 // See JBS-8090026 for more detail.
-                Platform.runLater(() -> Utils.removeMnemonics(popup, getSkinnable().getScene()));
+                Scene scene = getSkinnable().getScene();
```
`Utils.addMnemonics` has the same check, so in case of macOS `mnemonics`  list is always empty

aghaisas replied:
I was wondering why this code works.
Apparently, Utils.addMnemonics() adds to 'mnemonics' list only with same condition if (!com.sun.javafx.PlatformUtil.isMac()) { ...
Hence,  mnemonics list is empty in case platform is macOs.

vlaaad commented:
Thanks for review. Is there any time frame for when this will be available in early access releases?

kevinrushforth commented:
> Thanks for review. Is there any time frame for when this will be available in early access releases?

I'll integrate it today. @johanvos can comment as to when EA builds of openjfx14 might start.

vlaaad commented:
@kevinrushforth @johanvos I updated to latest (javafx 13, 13-ea+14b) and this fix is not included... When it will be released?

kevinrushforth commented:
If you click on the JBS issue, you will not that it is fixed in openjfx14.

@johanvos or @tiainen can comment as to when EA builds of openjfx14 will be available.

tiainen commented:
@vlaaad JavaFX 14-ea+1 has been made available.
